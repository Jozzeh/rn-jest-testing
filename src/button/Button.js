import React, {useState, useEffect} from 'react';

import PropTypes from 'prop-types';
import {styles} from './Button.style';
import {Text, TouchableHighlight} from 'react-native';

export const buttonTypes = {primary: 'primary', secondary: 'secondary'};

const availableButtonTypes = [buttonTypes.primary, buttonTypes.secondary];
const Button = (props) => {
  const [pressed, setPressed] = useState(false);

  useEffect(() => {
    standaloneFunctie();
  });

  function standaloneFunctie() {
    return true;
  }

  function buttonPressed() {
    return 'Testing is cool.';
  }

  return (
    <TouchableHighlight
      testID="buttonTester"
      style={[
        styles.lightMode.button.default,
        styles.lightMode.button.medium,
        styles.lightMode.button[props.type][pressed ? 'active' : 'default'],
      ]}
      activeOpacity={1}
      onHideUnderlay={() => {
        setPressed(false);
      }}
      onShowUnderlay={() => {
        setPressed(true);
      }}
      underlayColor={styles.lightMode.button[props.type].active.backgroundColor}
      onPress={() => {
        return buttonPressed();
      }}>
      <Text
        testID="buttonText"
        style={[
          styles.lightMode.title.default,
          styles.lightMode.title[props.size],
          styles.lightMode.title[props.type][pressed ? 'active' : 'default'],
        ]}>
        {props.title.toUpperCase()}
      </Text>
    </TouchableHighlight>
  );
};

Button.propTypes = {
  type: PropTypes.oneOf(availableButtonTypes),
  enabled: PropTypes.bool,
  title: PropTypes.string.isRequired,
};

Button.defaultProps = {
  title: 'Button',
  enabled: true,
  type: buttonTypes.primary,
};

export default Button;
