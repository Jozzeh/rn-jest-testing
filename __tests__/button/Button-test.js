/**
 * @format
 */

import 'react-native';
import React from 'react';
import Button from '../../src/button/Button';

// Note: test renderer must be required after react-native.
import renderer, {act} from 'react-test-renderer';

// *** RENDERING & SNAPSHOTTING ***
it('Button renders correctly', () => {
  const button = renderer.create(<Button />).toJSON();
  expect(button).toMatchSnapshot();
});

// *** RENDERING AND ASSERTING VALUE ***
it('Button title prop renders correctly', () => {
  const testValue = 'Test';
  const button = renderer.create(<Button title={testValue} />);
  let buttonobj = button.root;
  const renderedElement = buttonobj.find(
    (el) => el.props.testID === 'buttonText',
  );

  expect(renderedElement.props.children).toBe(testValue.toUpperCase());
});

// *** ASSERT RETURN VALUE OF EVENTS / FUNCTIONS ***
it('Button fires click event', () => {
  const button = renderer.create(<Button />);
  let instanceobj = button.root;
  const renderedElement = instanceobj.find(
    (el) => el.props.testID === 'buttonTester',
  );

  const pressValue = renderedElement.props.onPress();
  expect(pressValue).toBe('Testing is cool.');
});

// *** ASSERT RETURN VALUE OF EVENTS / FUNCTIONS ***
it('Button fire other prop events', () => {
  let button;
  act(() => {
    button = renderer.create(<Button />);
  });
  let instanceobj = button.root;
  const renderedElement = instanceobj.find(
    (el) => el.props.testID === 'buttonTester',
  );

  // here we check if the function runs...
  act(() => {
    renderedElement.props.onHideUnderlay();
    renderedElement.props.onShowUnderlay();
  });
});
