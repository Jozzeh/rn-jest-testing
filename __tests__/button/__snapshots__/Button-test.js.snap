// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 2,
        "borderWidth": 1.5,
        "justifyContent": "center",
      },
      Object {
        "height": 48,
      },
      Object {
        "backgroundColor": "#005BAA",
        "borderColor": "#005BAA",
      },
    ]
  }
  testID="buttonTester"
>
  <Text
    style={
      Array [
        undefined,
        undefined,
        Object {
          "color": "#fff",
        },
      ]
    }
    testID="buttonText"
  >
    BUTTON
  </Text>
</View>
`;
